modules/code-editor/src/views/full/FileNavigator.tsx

```diff
@@ -77,14 +83,87 @@ export default class FileNavigator extends React.Component<Props, State> {
     }
   }
 
+  handleContextMenu = (node: ITreeNode<NodeData>, path, e) => {
+    if (!node.nodeData) {
+      return null
+    }
+
+    e.preventDefault()
+    ContextMenu.show(
+      <Menu>
+        <MenuItem icon="edit" text="Rename" onClick={() => this.handleRename(node)} />
+        <MenuItem icon="delete" text="Delete" onClick={() => this.props.onNodeDelete(node.nodeData as EditableFile)} />
+      </Menu>,
+      { left: e.clientX, top: e.clientY }
+    )
+  }
+
+  handleRename = (node: ITreeNode<NodeData>) => {
+    const nodeDomElement = this.treeRef.current.getNodeContentElement(node.id)
+
+    const input = document.createElement('input')
+    input.type = 'text'
+    input.className = 'bp3-input bp3-small'
+    input.value = node.label as string
+
+    const div = document.createElement('div')
+    div.className = nodeDomElement.className
+    div.appendChild(input)
+
+    nodeDomElement.replaceWith(div)
+
+    input.focus()
+    input.select()
+
+    const closeRename = async e => {
+      e.preventDefault()
+      div.replaceWith(nodeDomElement)
+      window.removeEventListener('keydown', keyboardListener)
+      window.removeEventListener('mousedown', mouseListener)
+
+      const file = node.nodeData as EditableFile
+      let newName = input.value as string
+      newName = newName.endsWith('.js') ? newName : newName + '.js'
+
+      if (newName === node.label) {
+        return
+      }
+
+      try {
+        await this.props.onNodeRename(file, newName)
+      } catch (e) {
+        return
+      }
+
+      node.label = newName
+    }
+
+    const keyboardListener = async e => {
+      if (e.key === 'Enter' || e.key === 'Escape') {
+        await closeRename(e)
+      }
+    }
+
+    const mouseListener = async e => {
+      if (!div.contains(e.target)) {
+        await closeRename(e)
+      }
+    }
+
+    window.addEventListener('keydown', keyboardListener)
+    window.addEventListener('mousedown', mouseListener)
+  }
```

Comment:
Could you make this as a react component ? Theres a lot of stuff happening here. You could then use ReactDOM.render(RenameInput, myDiv)

Reply:
I actually did make a component and used ReactDOM.render but there is no gain in code clarity as I still need to create a React ref for each UI element (<div> and <input>)  and create a temporary div with document.createElement to render in.

The current solution is ,in my opinion, the best. Howerver, it is true that a lot of stuff happens on file rename, so I extracted this big piece of logic in tree.ts. The function signature is:

renameTreeNode(
  node: ITreeNode<any>,
  treeRef: React.RefObject<Tree<any>>,
  onRename: (newName: string) => Promise<void>
): Promise<void>